File: data_temp/merge_quran.dart
import 'dart:convert';
import 'dart:io';

void main() async {
  print("Loading datasets...");
  
  List<String> loadLines(String filepath) {
    var file = File(filepath);
    if (!file.existsSync()) {
      print("File not found: $filepath");
      return [];
    }
    return file.readAsLinesSync().where((line) {
      var t = line.trim();
      return t.isNotEmpty && !t.startsWith('#') && !t.startsWith('=');
    }).toList();
  }

  var arabicLines = loadLines('d:/Coding/Flutter/Muslim Launcher 2/data_temp/quran-tajweed.txt');
  var latinLines = loadLines('d:/Coding/Flutter/Muslim Launcher 2/data_temp/en.transliteration.txt');
  var idTransLines = loadLines('d:/Coding/Flutter/Muslim Launcher 2/data_temp/id.indonesian.txt');
  var enTransLines = loadLines('d:/Coding/Flutter/Muslim Launcher 2/data_temp/en.sahih.txt');

  print("Lines found: Arabic(${arabicLines.length}), Latin(${latinLines.length}), ID Trans(${idTransLines.length}), EN Trans(${enTransLines.length})");

  print("Extracting metadata from existing quran.json...");
  var surahMeta = <int, Map<String, dynamic>>{};
  var quranFile = File('d:/Coding/Flutter/Muslim Launcher 2/assets/quran.json');
  if (quranFile.existsSync()) {
    var content = quranFile.readAsStringSync();
    var existingQuran = json.decode(content) as List;
    for (var surah in existingQuran) {
      var num = surah['surah_number'] as int;
      surahMeta[num] = {
        'surah_name': surah['surah_name'],
        'total_ayah': surah['total_ayah'],
      };
    }
  }

  print("Merging data into new structure...");
  var newQuran = [];
  
  int lineIndex = 0;

  for (var surahNum = 1; surahNum <= 114; surahNum++) {
    var surahInfo = {
      "surah_number": surahNum,
      "surah_name": surahMeta[surahNum]?['surah_name'] ?? "Surah $surahNum",
      "total_ayah": surahMeta[surahNum]?['total_ayah'] ?? 0,
      "ayahs": []
    };

    var totalAyah = surahInfo["total_ayah"] as int;
    for (var ayahNum = 1; ayahNum <= totalAyah; ayahNum++) {
      String arabic = (lineIndex < arabicLines.length) ? arabicLines[lineIndex] : "";
      
      surahInfo["ayahs"].add({
        "ayah_number": ayahNum,
        "arabic": arabic,
        "latin": (lineIndex < latinLines.length) ? latinLines[lineIndex] : "",
        "translation_id": (lineIndex < idTransLines.length) ? idTransLines[lineIndex] : "",
        "translation_en": (lineIndex < enTransLines.length) ? enTransLines[lineIndex] : ""
      });
      lineIndex++;
    }
    newQuran.add(surahInfo);
  }

  print("Total processed lines: $lineIndex");

  print("Writing to assets/quran.json...");
  var encoder = JsonEncoder.withIndent(null); // Compact formatting
  var encodedJson = encoder.convert(newQuran);
  quranFile.writeAsStringSync(encodedJson);
  
  print("Done! File size: ${quranFile.lengthSync()} bytes.");
}

File: data_temp/merge_quran.py
import json
import os

def load_data(filepath):
    """Loads a text file where each line is 'surah|ayah|text'."""
    data = {}
    if not os.path.exists(filepath):
        print(f"File not found: {filepath}")
        return data
        
    with open(filepath, 'r', encoding='utf-8') as f:
        for line in f:
            parts = line.strip().split('|', 2)
            if len(parts) >= 3:
                surah = int(parts[0])
                ayah = int(parts[1])
                text = parts[2]
                
                if surah not in data:
                    data[surah] = {}
                data[surah][ayah] = text
    return data

def main():
    print("Loading datasets...")
    # Load all datasets
    arabic_data = load_data('data_temp/quran-uthmani.txt')
    latin_data = load_data('data_temp/en.transliteration.txt')
    id_trans_data = load_data('data_temp/id.indonesian.txt')
    en_trans_data = load_data('data_temp/en.sahih.txt')
    
    # We also need surah names and ayah counts. 
    # The existing quran.json has this, so let's extract metadata from it first.
    print("Extracting metadata from existing quran.json...")
    surah_meta = {}
    try:
        with open('assets/quran.json', 'r', encoding='utf-8') as f:
            existing_quran = json.load(f)
            for surah in existing_quran:
                surah_num = surah['surah_number']
                surah_meta[surah_num] = {
                    'surah_name': surah['surah_name'],
                    'total_ayah': surah['total_ayah']
                }
    except Exception as e:
        print(f"Error reading existing quran.json: {e}")
        return

    print("Merging data into new structure...")
    new_quran = []
    
    for surah_num in range(1, 115):
        if surah_num not in surah_meta:
            continue
            
        surah_info = {
            "surah_number": surah_num,
            "surah_name": surah_meta[surah_num]['surah_name'],
            "total_ayah": surah_meta[surah_num]['total_ayah'],
            "ayahs": []
        }
        
        for ayah_num in range(1, surah_meta[surah_num]['total_ayah'] + 1):
            ayah_info = {
                "ayah_number": ayah_num,
                "arabic": arabic_data.get(surah_num, {}).get(ayah_num, ""),
                "latin": latin_data.get(surah_num, {}).get(ayah_num, ""),
                "translation_id": id_trans_data.get(surah_num, {}).get(ayah_num, ""),
                "translation_en": en_trans_data.get(surah_num, {}).get(ayah_num, "")
            }
            surah_info["ayahs"].append(ayah_info)
            
        new_quran.append(surah_info)
        
    print("Writing to assets/quran.json...")
    with open('assets/quran.json', 'w', encoding='utf-8') as f:
        json.dump(new_quran, f, ensure_ascii=False, separators=(',', ':'))
        
    print("Done! File size:", os.path.getsize('assets/quran.json'))

if __name__ == '__main__':
    main()
